cps/kni.c

```diff
@@ -850,6 +850,273 @@ rd_send_err(const struct nlmsghdr *req, int err)
 	}
 }
 
+static void
+rd_fill_getroute_reply(struct cps_config *cps_conf, struct nlmsghdr *reply,
+	struct gk_fib *fib, int family, uint32_t seq,
+	uint8_t prefix_len, struct ipaddr **gw_addr)
+{
+	struct rtmsg *rm;
+
+	reply->nlmsg_type = RTM_NEWROUTE;
+	reply->nlmsg_flags = NLM_F_MULTI;
+	reply->nlmsg_seq = seq;
+	reply->nlmsg_pid = cps_conf->nl_pid;
+
+	rm = mnl_nlmsg_put_extra_header(reply, sizeof(*rm));
+	rm->rtm_family = family;
+	rm->rtm_dst_len = prefix_len;
+	rm->rtm_src_len = 0;
+	rm->rtm_tos = 0;
+	rm->rtm_table = RT_TABLE_MAIN;
+	rm->rtm_scope = RT_SCOPE_UNIVERSE;
+	rm->rtm_type = RTN_UNICAST;
+	rm->rtm_flags = 0;
+
+	switch (fib->action) {
+	case GK_FWD_GRANTOR:
+		mnl_attr_put_u32(reply, RTA_OIF, cps_conf->back_kni_index);
+		rm->rtm_protocol = RTPROT_STATIC;
+		*gw_addr = &fib->u.grantor.eth_cache->ip_addr;
+		break;
+	case GK_FWD_GATEWAY_FRONT_NET:
+		mnl_attr_put_u32(reply, RTA_OIF, cps_conf->front_kni_index);
+		rm->rtm_protocol = fib->u.gateway.rt_proto;
+		*gw_addr = &fib->u.gateway.eth_cache->ip_addr;
+		break;
+	case GK_FWD_GATEWAY_BACK_NET:
+		mnl_attr_put_u32(reply, RTA_OIF, cps_conf->back_kni_index);
+		rm->rtm_protocol = fib->u.gateway.rt_proto;
+		*gw_addr = &fib->u.gateway.eth_cache->ip_addr;
+		break;
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		mnl_attr_put_u32(reply, RTA_OIF, cps_conf->front_kni_index);
+		rm->rtm_protocol = RTPROT_UNSPEC;
+		*gw_addr = NULL;
+		break;
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		mnl_attr_put_u32(reply, RTA_OIF, cps_conf->back_kni_index);
+		rm->rtm_protocol = RTPROT_UNSPEC;
+		*gw_addr = NULL;
+		break;
+	default:
+		rte_panic("Invalid FIB action (%u) in FIB while being processed by CPS block in %s\n",
+			fib->action, __func__);
+		return;
+	}
+}
+
+static int
+rd_send_batch(struct cps_config *cps_conf, struct mnl_nlmsg_batch *batch,
+	const char *daemon, uint32_t seq, uint32_t pid, int done)
+{
+	/* Address of routing daemon. */
+	struct sockaddr_nl rd_sa;
+	size_t len = mnl_nlmsg_batch_size(batch);
```

Comment:
We don't need variable `len`. It's true that `len` is updated in the body of the `if` statement, but `batch` is also updated there when `mnl_nlmsg_put_header()` is called. Once that update goes away, `len` is used only once.